### PR TITLE
feat: attach/detach bound devices in tray context menu

### PR DESCRIPTION
**Pull request by JPHutchins**
This PR adds a "nice to have" feature.  Generally, I think that the GUI is great and it would be a mistake to put too much functionality in the system tray. Instead, users should be encourage dto open the main window where operations will be less ambiguous.

However, my day-to-day usage rarely involves inspecting unbound devices, binding devices, or setting up auto attach. I have 2 or 3 bound devices and I either want them attached or not.

And so, this PR adds the connected + bound devices in the context menu.  A check mark is displayed next to devices that are forwarded. Clicking on a device either attaches or detaches depending on the state.

image

I am new to Rust and this was hard for me to do. 😢 I am concerned about memory leaks. I have already resolved a stack overflow caused by the event handler closure not being unbound. It was getting added to the stack on every context menu open:

```rust
if let Some(handler) = self.menu_tray_event_handler.borrow().as_ref() {
    nwg::unbind_event_handler(handler);
}
```
I've tested the current implementation quite a bit now and it seems OK, but it could be totally wrong and unsafe - I simply don't have enough experience with Rust to know! Nevertheless, I think the feature is a good one.

Unaddressed issues:
* No handling of devices that are auto attach. Ideally they would be greyed out since attaching/detaching is pointless.
* Windows annoying lag during USB enumeration. I've seen this before, it's not a Rust problem. During attach, the context menu will become unresponsive, presumably because the thread is "busy" doing the attach. Would be nice to let the user know what's happening somehow but IMO it's not too bad.

### CODE REVIEW

File: src/gui/usbipd_gui.rs [hunk 1]
```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
         self.window.set_visible(false);
     }
 
-    fn show(&self) {
+    fn show(self: &Rc<UsbipdGui>) {
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:140]: Any reason for this change? It doesn't look like an `Rc` is needed in this method.

File: src/gui/usbipd_gui.rs [hunk 2]
```diff
@@ -25,6 +26,7 @@ pub(super) trait GuiTab {
 #[derive(Default, NwgUi)]
 pub struct UsbipdGui {
     device_notification: Cell<DeviceNotification>,
+    menu_tray_event_handler: RefCell<Option<nwg::EventHandler>>,
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:29]: You could use `Cell` instead of `RefCell` to avoid having to borrow stuff around, since you only need to `Cell:set` when creating the event handler, and either use `Cell:take` or `Cell:update` when unbinding the handler.

File: src/gui/usbipd_gui.rs [hunk 3]
```diff
@@ -75,37 +77,22 @@ pub struct UsbipdGui {
...
-    #[nwg_events(OnMenuItemSelected: [UsbipdGui::exit])]
-    menu_tray_exit: nwg::MenuItem,
-
     // File menu
     #[nwg_control(parent: window, text: "File", popup: false)]
     menu_file: nwg::Menu,
 
     #[nwg_control(parent: menu_file, text: "Refresh")]
-    #[nwg_events(OnMenuItemSelected: [UsbipdGui::refresh])]
+    #[nwg_events(OnMenuItemSelected: [UsbipdGui::refresh(RC_SELF)])]
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:88]: Revert, not needed.

File: src/gui/usbipd_gui.rs [hunk 4]
```diff
@@ -75,37 +77,22 @@ pub struct UsbipdGui {
...
     menu_file_refresh: nwg::MenuItem,
 
     #[nwg_control(parent: menu_file)]
     menu_file_sep1: nwg::MenuSeparator,
 
     #[nwg_control(parent: menu_file, text: "Exit")]
-    #[nwg_events(OnMenuItemSelected: [UsbipdGui::exit])]
+    #[nwg_events(OnMenuItemSelected: [UsbipdGui::exit()])]
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:95]: Revert, not needed.

File: src/gui/usbipd_gui.rs [hunk 5]
```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
...
+    ) -> Result<nwg::MenuSeparator, nwg::NwgError> {
+        let mut sep = nwg::MenuSeparator::default();
+        nwg::MenuSeparator::builder()
+            .parent(parent)
+            .build(&mut sep)
+            .map(|_| sep)
     }
 
-    fn refresh(&self) {
+    fn refresh(self: &Rc<UsbipdGui>) {
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:247]: Revert, not needed.

File: src/gui/usbipd_gui.rs [hunk 6]
```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
...
         self.connected_tab_content.refresh();
         self.persisted_tab_content.refresh();
         self.auto_attach_tab_content.refresh();
     }
 
-    fn exit(&self) {
+    fn exit() {
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:253]: Revert, not needed.

File: src/usbipd.rs [hunk 7]
```diff
@@ -47,7 +47,7 @@ impl Display for UsbipState {
 }
 
 /// A struct representing a USB device as returned by `usbipd`.
-#[derive(Debug, Deserialize)]
+#[derive(Debug, Deserialize, Clone)]
```
Comment (nickbeth) [inline src/usbipd.rs:50]: Shouldn't be needed if you follow the other comment, remove it.

File: src/gui/usbipd_gui.rs [hunk 8]
```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
...
+                                        if device.is_attached() {
+                                            device.detach().ok();
+                                        } else {
+                                            device.attach().ok();
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        _ => (),
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:211]: You should add a handler case for "menu closed" (don't know the name of the event) that unbinds the event handler here instead of where you do it at the beginning of this function.

**Review by nickbeth — CHANGES_REQUESTED** [review]
This is a good change so I'll accept it. However there's some changes needed before I merge this.

File: src/gui/usbipd_gui.rs [hunk 9]
```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
...
+            .build(&mut menu_tray)
+            .unwrap();
+
+        let devices = self
+            .connected_tab_content
+            .connected_devices
+            .borrow()
+            .iter()
+            .cloned()
+            .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (JPHutchins) [reply src/gui/usbipd_gui.rs:162]: Aw, that's much better!

File: src/gui/usbipd_gui.rs (hunk 2 above)
Comment (JPHutchins) [reply src/gui/usbipd_gui.rs:29]: This is great, I'd love to avoid RefCell!

File: src/gui/usbipd_gui.rs (hunk 4 above)
Comment (JPHutchins) [reply src/gui/usbipd_gui.rs:95]: It's not needed, but exit() does not need a reference to self, so why expose it?

File: src/gui/usbipd_gui.rs (hunk 8 above)
Comment (JPHutchins) [reply src/gui/usbipd_gui.rs:211]: That will be much better! This was quite ugly to deal with.

Comment (JPHutchins) [reply src/gui/usbipd_gui.rs]: I tried the menu_exit, and unfortunately it caused the other handlers to not work. When they are clicked on, an exit event occurs, so it's either that:

* exit event is handled before the selected event (handler gets unregistered first)
* or they are handled in different threads causing a race condition

I could look into the implementation to understand.

I've left the unbind_event_handler in the menu open for now.

File: src/gui/usbipd_gui.rs [hunk 10]
```diff
@@ -154,9 +141,102 @@ impl UsbipdGui {
...
+            if device.is_bound() {
+                let menu_item = self
+                    .new_menu_item(menu_tray.handle, &description, device.is_attached())
+                    .unwrap();
+
+                menu_items.push((menu_item, device));
+            }
+        }
+
+        self.new_menu_separator(menu_tray.handle).unwrap();
```
Comment (nickbeth) [inline src/gui/usbipd_gui.rs:180]: Ideally this separator should only show when at least one device is displayed in the menu, otherwise it wouldn't look good. Alternatively, we can show a greyed-out (disabled) option saying something like "No bound devices" to hint the user that bound devices will be shown there. Let me know what you think.

**Review by nickbeth — COMMENTED** [review]
So I've actually gone and tested this PR, it's a really neat feature!

There's one more issue though: errors while attaching are not reported to the user. We should really spawn a dialog box with the error message, see how it's done in other parts of the attaching code.

Additionally, I've noticed that since attaching takes quite some time, once you click on a device in the menu for attaching it, the menu disappears and it won't open up until the attaching is completed. Now, I'm not even sure how to easily fix this, as every way I come up with requires complex state tracking, so maybe it's fine?

File: src/gui/usbipd_gui.rs (hunk 8 above)
Comment (nickbeth) [reply src/gui/usbipd_gui.rs]: Right, I'm fine with leaving as is, as long as you add a comment in the code explaining what you said here, for documentation purposes. Something along the lines of 
> Unbinding of the menu handler is done here because the exit event is received before menu action events, therefore unbinding on the exit event would cause other actions to not be handled.

**Comment by JPHutchins** [comment]
> There's one more issue though: errors while attaching are not reported to the user. We should really spawn a dialog box with the error message, see how it's done in other parts of the attaching code.

Agreed! I think that I can cause an error by opening the menu, unplugging the device, then attempting to forward it. And trying to forward with WSL shutdown.

**Comment by JPHutchins** [comment]
> Additionally, I've noticed that since attaching takes quite some time, once you click on a device in the menu for attaching it, the menu disappears and it won't open up until the attaching is completed. Now, I'm not even sure how to easily fix this, as every way I come up with requires complex state tracking, so maybe it's fine?

Yeah, it's the same for the main window. I hate threads and state, so my vote is to keep it simple. I see this whenever interacting with Windows USB. Their APIs must block for 200ms - 2s during attach/detach events.

File: src/gui/usbipd_gui.rs (hunk 10 above)
Comment (JPHutchins) [reply src/gui/usbipd_gui.rs:180]: I like the greyed-out "No bound devices" - simpler implementation and self-documenting.

**Comment by JPHutchins** [comment]
I've been running this branch from morning to evening everyday at work for the last month - haven't noticed any problems.

**Comment by nickbeth** [comment]
Sorry for the long absence time, I've finally had time to properly look at the final version of this. I've fixed up a bunch of things, mainly just further reduced reduced the indentation level, and inserted a TODO comment for the UI hang I've mentioned previously. For now that shouldn't be a big issue though.

Thanks again for this PR, I'm going to finally merge it now, and maybe make a new release since this feature is worth one.

**Review by nickbeth — APPROVED** [review]
